x/gov/proposal.go

```diff
@@ -0,0 +1,182 @@
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get(ProposalKey(proposalID))
+	if bz == nil {
+		return
+	}
+	keeper.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &proposal)
+	return proposal, true
+}
+
+// SetProposal set a proposal to store
+func (keeper Keeper) SetProposal(ctx sdk.Context, proposal Proposal) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinaryLengthPrefixed(proposal)
+	store.Set(ProposalKey(proposal.ProposalID), bz)
+}
+
+// DeleteProposal deletes a proposal from store
+func (keeper Keeper) DeleteProposal(ctx sdk.Context, proposalID uint64) {
+	store := ctx.KVStore(keeper.storeKey)
+	proposal, ok := keeper.GetProposal(ctx, proposalID)
+	if !ok {
+		panic("DeleteProposal cannot fail to GetProposal")
+	}
+	keeper.RemoveFromInactiveProposalQueue(ctx, proposalID, proposal.DepositEndTime)
+	keeper.RemoveFromActiveProposalQueue(ctx, proposalID, proposal.VotingEndTime)
+	store.Delete(ProposalKey(proposalID))
+}
+
+// GetProposals returns all the proposals from store
+func (keeper Keeper) GetProposals(ctx sdk.Context) (proposals Proposals) {
+	keeper.IterateProposals(ctx, func(proposal types.Proposal) bool {
+		proposals = append(proposals, proposal)
+		return false
+	})
+	return
+}
+
+// GetProposalsFiltered get Proposals from store by ProposalID
+// voterAddr will filter proposals by whether or not that address has voted on them
+// depositorAddr will filter proposals by whether or not that address has deposited to them
+// status will filter proposals by status
+// numLatest will fetch a specified number of the most recent proposals, or 0 for all proposals
+func (keeper Keeper) GetProposalsFiltered(ctx sdk.Context, voterAddr sdk.AccAddress, depositorAddr sdk.AccAddress, status ProposalStatus, numLatest uint64) []Proposal {
+
+	maxProposalID, err := keeper.peekCurrentProposalID(ctx)
+	if err != nil {
+		return nil
+	}
+
+	matchingProposals := []Proposal{}
+
+	if numLatest == 0 {
+		numLatest = maxProposalID
+	}
+
+	for proposalID := maxProposalID - numLatest; proposalID < maxProposalID; proposalID++ {
+		if voterAddr != nil && len(voterAddr) != 0 {
+			_, found := keeper.GetVote(ctx, proposalID, voterAddr)
+			if !found {
+				continue
+			}
+		}
+
+		if depositorAddr != nil && len(depositorAddr) != 0 {
+			_, found := keeper.GetDeposit(ctx, proposalID, depositorAddr)
+			if !found {
+				continue
+			}
+		}
+
+		proposal, ok := keeper.GetProposal(ctx, proposalID)
+		if !ok {
+			continue
+		}
+
+		if ValidProposalStatus(status) {
+			if proposal.Status != status {
+				continue
+			}
+		}
+
+		matchingProposals = append(matchingProposals, proposal)
+	}
+	return matchingProposals
+}
+
+// GetLastProposalID gets the last used proposal ID
+func (keeper Keeper) GetLastProposalID(ctx sdk.Context) (proposalID uint64) {
+	proposalID, err := keeper.peekCurrentProposalID(ctx)
+	if err != nil {
+		return 0
+	}
+	proposalID--
```

Comment:
What if the current proposal ID is 0?

Reply:
> What if the current proposal ID is 0?

good catch. More generally, this could even make `proposalID < minProposalID` (in the case where `minProposalID` > 1)

Reply:
I was checking and there's no way to get the starting proposal ID, it's set to the NextProposalID and then incremented every time a new proposal is created. I guess we can leave as follows: 

```
	proposalID, err := keeper.peekCurrentProposalID(ctx)
	if err != nil || proposalID == 0 {
		return 0
	}
	return proposalID - 1
```

Reply:
updated as part of Karoly's suggestions